In [1]:
# import packages
import pandas as pd
import numpy as np

## Part 1

### Question 1a

In [13]:
patterns_2018 = pd.read_csv('weekly_patterns_2018_sample.csv.zip',
                            compression = 'zip')
patterns_2019 = pd.read_csv('weekly_patterns_2019_sample.csv.zip',
                            compression = 'zip')
patterns_2020 = pd.read_csv('weekly_patterns_2020_sample.csv.zip',
                            compression = 'zip')
patterns_2021 = pd.read_csv('weekly_patterns_2021_sample.csv.zip',
                            compression = 'zip')
patterns_2022 = pd.read_csv('weekly_patterns_2022_sample.csv.zip',
                            compression = 'zip')
total_obsv = (
    len(patterns_2018) +
    len(patterns_2019) +
    len(patterns_2020) +
    len(patterns_2021) +
    len(patterns_2022)
)
print(total_obsv)

6462235


Total number of observations in all datasets = 6,462,235
Unit of observation = Weekly visits by an individual to specific location

### Question 1b

In [21]:
# Filtering each dataset for rows where the 'brands' = 'Zaxby's'
filtered_2018 = patterns_2018[patterns_2018['brands'] == "Zaxby's"]
filtered_2019 = patterns_2019[patterns_2019['brands'] == "Zaxby's"]
filtered_2020 = patterns_2020[patterns_2020['brands'] == "Zaxby's"]
filtered_2021 = patterns_2021[patterns_2021['brands'] == "Zaxby's"]
filtered_2022 = patterns_2022[patterns_2022['brands'] == "Zaxby's"]

# Concatenate the filtered datasets
df_conc = pd.concat([filtered_2018, filtered_2019, 
                             filtered_2020, filtered_2021, 
                             filtered_2022], ignore_index=True)

# Display the total number of observations in the concatenated dataframe
print(f"Total observations for Zaxby's across all years: {len(df_conc)}")
df_conc.head(3)

Total observations for Zaxby's across all years: 44373


placekey          city region           date_range_start  \
0  223-222@8fx-zfv-9cq     Melbourne     FL  2018-04-30T00:00:00-04:00   
1  223-222@5ps-jj2-qj9  Jacksonville     AR  2018-10-22T00:00:00-05:00   
2  zzw-222@8gf-bw8-rff        Shelby     NC  2018-06-11T00:00:00-04:00   

              date_range_end  raw_visit_counts           visits_by_day  \
0  2018-05-07T00:00:00-04:00               207  [24,37,32,24,21,43,26]   
1  2018-10-29T00:00:00-05:00                89     [6,8,20,13,16,17,9]   
2  2018-06-18T00:00:00-04:00               123  [16,15,10,15,26,24,17]   

                         safegraph_brand_ids  naics_code  postal_code   brands  
0  SG_BRAND_5271fc9e8c38fe53392a2ecabc642130      722513        32940  Zaxby's  
1  SG_BRAND_5271fc9e8c38fe53392a2ecabc642130      722513        72076  Zaxby's  
2  SG_BRAND_5271fc9e8c38fe53392a2ecabc642130      722513        28152  Zaxby's

### Question 1c

What is the data type of each variable in your wide sample?

In [32]:
df_conc.dtypes

placekey               object
city                   object
region                 object
date_range_start       object
date_range_end         object
raw_visit_counts        int64
visits_by_day          object
safegraph_brand_ids    object
naics_code              int64
postal_code             int64
brands                 object
dtype: object

How many unique places do you observe for your chain in the US?

In [37]:
unique_places = df_conc['placekey'].nunique()
print(f"Number of unique Zaxby's locations in the dataset: {unique_places}")

Number of unique Zaxby's locations in the dataset: 180


The Official Zaxby's website mentions that they have >900 locations.
The discrepancy in numbers could be caused by data collection timeframes, 
recently opened/closed locations, and data availability. 

In [36]:
df_conc.to_csv('p1q1_CONC.csv',
         index=False)

## Part 2

### Question 2a

In [51]:
# Convert date_range_start and date_range_end to datetime format, handling 
# utc error given before
df_conc['date_range_start'] = pd.to_datetime(df_conc['date_range_start'], utc=True)
df_conc['date_range_end'] = pd.to_datetime(df_conc['date_range_end'], utc=True)

# Find the beginning and ending dates (staying in datetime format to 
# keep UTC conversion accurate)
beginning_date = df_conc['date_range_start'].min()
ending_date = df_conc['date_range_end'].max()

# Print beginning and ending dates
print(f"Beginning date in the dataset: {beginning_date}")
print(f"Ending date in the dataset: {ending_date}")


Beginning date in the dataset: 2018-01-01 05:00:00+00:00
Ending date in the dataset: 2023-01-02 07:00:00+00:00


### Question2b

In [57]:
print(df_conc['visits_by_day'].head())

0    [24, 37, 32, 24, 21, 43, 26]
1       [6, 8, 20, 13, 16, 17, 9]
2    [16, 15, 10, 15, 26, 24, 17]
3     [9, 16, 10, 14, 27, 15, 14]
4    [12, 22, 23, 19, 34, 13, 17]
Name: visits_by_day, dtype: object


In [63]:
# For loop to create new columns for each day's visits directly from the lists
for i in range(1, 8):  # 1 through 7 for dailyvisits1 through dailyvisits7
    df_conc[f'dailyvisits{i}'] = (df_conc['visits_by_day'].apply
                                  (lambda x: x[i-1] if x is not None else None))

# Count the total number of variables (columns) in df_conc now
total_variables = len(df_conc.columns)
print(f"Total number of variables in the 'wide sample' now: {total_variables}")
df_conc.head()

Total number of variables in the 'wide sample' now: 18


placekey          city region          date_range_start  \
0  223-222@8fx-zfv-9cq     Melbourne     FL 2018-04-30 04:00:00+00:00   
1  223-222@5ps-jj2-qj9  Jacksonville     AR 2018-10-22 05:00:00+00:00   
2  zzw-222@8gf-bw8-rff        Shelby     NC 2018-06-11 04:00:00+00:00   
3  222-222@8gg-zrq-pd9        Camden     SC 2018-04-02 04:00:00+00:00   
4  223-222@8gq-4wb-5s5         Ocala     FL 2018-09-03 04:00:00+00:00   

             date_range_end  raw_visit_counts                 visits_by_day  \
0 2018-05-07 04:00:00+00:00               207  [24, 37, 32, 24, 21, 43, 26]   
1 2018-10-29 05:00:00+00:00                89     [6, 8, 20, 13, 16, 17, 9]   
2 2018-06-18 04:00:00+00:00               123  [16, 15, 10, 15, 26, 24, 17]   
3 2018-04-09 04:00:00+00:00               105   [9, 16, 10, 14, 27, 15, 14]   
4 2018-09-10 04:00:00+00:00               140  [12, 22, 23, 19, 34, 13, 17]   

                         safegraph_brand_ids  naics_code  postal_code  \
0  SG_BRAND_5271fc9e8c38fe53392a2ecabc642130      722513        32940   
1  SG_BRAND_5271fc9e8c38fe53392a2ecabc642130      722513        72076   
2  SG_BRAND_5271fc9e8c38fe53392a2ecabc642130      722513        28152   
3  SG_BRAND_5271fc9e8c38fe53392a2ecabc642130      722513        29020   
4  SG_BRAND_5271fc9e8c38fe53392a2ecabc642130      722513        34470   

    brands  dailyvisits1  dailyvisits2  dailyvisits3  dailyvisits4  \
0  Zaxby's            24            37            32            24   
1  Zaxby's             6             8            20            13   
2  Zaxby's            16            15            10            15   
3  Zaxby's             9            16            10            14   
4  Zaxby's            12            22            23            19   

   dailyvisits5  dailyvisits6  dailyvisits7  
0            21            43            26  
1            16            17             9  
2            26            24            17  
3            27            15            14  
4            34            13            17

### Question 2c

In [73]:
# Melting df_conc to create the long DataFrame
df_long = (pd.melt(df_conc, id_vars=[col for col in df_conc.columns 
                  if not col.startswith('dailyvisits')],
                  value_vars=[f'dailyvisits{i}' for i in range(1, 8)],
                  var_name='day', value_name='dailyvisits'))

# Convert 'day' to just the numeric day of the week
df_long['day'] = df_long['day'].str.extract('(\d+)').astype(int)

# Display the first few rows to verify the transformation
df_long.head()


placekey          city region          date_range_start  \
0  223-222@8fx-zfv-9cq     Melbourne     FL 2018-04-30 04:00:00+00:00   
1  223-222@5ps-jj2-qj9  Jacksonville     AR 2018-10-22 05:00:00+00:00   
2  zzw-222@8gf-bw8-rff        Shelby     NC 2018-06-11 04:00:00+00:00   
3  222-222@8gg-zrq-pd9        Camden     SC 2018-04-02 04:00:00+00:00   
4  223-222@8gq-4wb-5s5         Ocala     FL 2018-09-03 04:00:00+00:00   

             date_range_end  raw_visit_counts                 visits_by_day  \
0 2018-05-07 04:00:00+00:00               207  [24, 37, 32, 24, 21, 43, 26]   
1 2018-10-29 05:00:00+00:00                89     [6, 8, 20, 13, 16, 17, 9]   
2 2018-06-18 04:00:00+00:00               123  [16, 15, 10, 15, 26, 24, 17]   
3 2018-04-09 04:00:00+00:00               105   [9, 16, 10, 14, 27, 15, 14]   
4 2018-09-10 04:00:00+00:00               140  [12, 22, 23, 19, 34, 13, 17]   

                         safegraph_brand_ids  naics_code  postal_code  \
0  SG_BRAND_5271fc9e8c38fe53392a2ecabc642130      722513        32940   
1  SG_BRAND_5271fc9e8c38fe53392a2ecabc642130      722513        72076   
2  SG_BRAND_5271fc9e8c38fe53392a2ecabc642130      722513        28152   
3  SG_BRAND_5271fc9e8c38fe53392a2ecabc642130      722513        29020   
4  SG_BRAND_5271fc9e8c38fe53392a2ecabc642130      722513        34470   

    brands  id  day  dailyvisits  
0  Zaxby's   0    1           24  
1  Zaxby's   1    1            6  
2  Zaxby's   2    1           16  
3  Zaxby's   3    1            9  
4  Zaxby's   4    1           12

How many observations do you have in the "long sample"?

In [74]:
print(f"Number of observations in the 'long sample': {len(df_long)}")

Number of observations in the 'long sample': 310611


A "long sample" allows for more flexibility in analysis. Each row contains a single observation, which make it easier to filter, group, and aggregate data. Long samples tend to take up more space which affects storage and processing times. 

In [78]:
df_long['dailyvisits'] = df_long['dailyvisits'].astype(int)
print(df_long['dailyvisits'])

0         24
1          6
2         16
3          9
4         12
          ..
310606    28
310607    12
310608    59
310609    23
310610    18
Name: dailyvisits, Length: 310611, dtype: int64


It worked! Which means there were no missing values(NaN), and no non-numeric strings.

### Question 2d

In [80]:
# Calculate descriptive statistics for 'dailyvisits'
description = df_long['dailyvisits'].describe()
print(description)

# To talk about outliers, I'm going to use Interquartile(IQR) range for clarity
# An outlier for our purposes is defined as a value that is 
# more than 1.5 * IQR below the first quartile (Q1) or above 
# the third quartile (Q3). I have also assigned Q1 and Q3.
Q1 = df_long['dailyvisits'].quantile(0.25)
Q3 = df_long['dailyvisits'].quantile(0.75)
IQR = Q3 - Q1

# Identify outliers
outliers = (df_long[(df_long['dailyvisits'] < (Q1 - 1.5 * IQR)) 
                | (df_long['dailyvisits'] > (Q3 + 1.5 * IQR))])
num_outliers = len(outliers)

# Percentage of outliers
percentage_outliers = (num_outliers / len(df_long)) * 100

print(f"Number of outliers: {num_outliers}")
print(f"Percentage of data considered outliers: {percentage_outliers:.2f}%")


count    310611.000000
mean         28.828139
std          22.802622
min           0.000000
25%          15.000000
50%          24.000000
75%          38.000000
max        1778.000000
Name: dailyvisits, dtype: float64
Number of outliers: 10703
Percentage of data considered outliers: 3.45%


In [81]:
# Count missing values
missing_values = df_long['dailyvisits'].isnull().sum()
percentage_missing = (missing_values / len(df_long)) * 100

print(f"Missing values in 'dailyvisits': {missing_values}")
print(f"Percentage of missing values: {percentage_missing:.2f}%")


Missing values in 'dailyvisits': 0
Percentage of missing values: 0.00%


#### Comments on missing values.
I did nto observe any missing values in dailyvisits as shows above. But, in the case that I did, some potential reasons for it could be stores being closed for renovations or ownership changes, and seasonal variability. Some days, perhaps when it's raining really hard or everyone is snowed in, may cause extremly low foot traffic to Zaxby's. 

#### How would outliers/missing values affect the quality of the dataset when you , as a business manager of the chain, attempt to make businesss decisions based on this dataset?
Outliers can skew my analyis.They can affect what I think my average visits are, and not account for event-specifc occurence. Missing values can introduce bias. Decisions that are immune to this would be those based on long term trends; such as overal strategic direction  year-on-year growth assessments. 

## Part 3 
### Question 3a

In [85]:
df_long['date_range_start'] = pd.to_datetime(df_long['date_range_start'])

# Create a 'date' variable
df_long['date'] = (df_long['date_range_start'] + 
                   pd.to_timedelta(df_long['day'] - 1,
                   unit='d'))

# Aggregating by year
df_long['year'] = df_long['date'].dt.year

yearly_summary = df_long.groupby('year').agg(
    start_date=('date', 'min'),
    end_date=('date', 'max'),
    observations=('dailyvisits', 'size'),
    mean_dailyvisits=('dailyvisits', 'mean')
).reset_index()

# stylizing table
styled_table = yearly_summary.style.set_table_styles(
    [{'selector': 'tr:hover',
      'props': [('background-color', '#ffff99')]}]  # Highlight on hover(my favorite)
).set_caption("Yearly Summary of Daily Visits")  # Title
styled_table



### Question 3b

In [108]:
df_long.reset_index(drop=True, inplace=True)
df_long['date_range_start'] = pd.to_datetime(df_long['date_range_start'])

# Ensure 'day' is zero-based for the calculation if day 1 means the start date
df_long['date'] = (df_long['date_range_start'] + 
                   pd.to_timedelta(df_long['day'] - 1, unit='D'))

df_long['dayofweek'] = df_long['date'].dt.day_name()

day_mapping = {
    'Monday': 1,
    'Tuesday': 2,
    'Wednesday': 3,
    'Thursday': 4,
    'Friday': 5,
    'Saturday': 6,
    'Sunday': 7
}

# Apply the mapping to create a numeric 'dayofweek_num' column
df_long['dayofweek_num'] = df_long['dayofweek'].map(day_mapping)

dayofweek_summary = df_long.groupby(['dayofweek', 'dayofweek_num']).agg(
    observations=('dailyvisits', 'size'),
    mean_dailyvisits=('dailyvisits', 'mean')
).reset_index().sort_values(by='dayofweek_num')

# Now that the aggregation is sorted, you can drop the numeric column for display
dayofweek_summary.drop('dayofweek_num', axis=1, inplace=True)

print(dayofweek_summary)

   dayofweek  observations  mean_dailyvisits
1     Monday         44373         26.011268
5    Tuesday         44373         26.741284
6  Wednesday         44373         27.905032
4   Thursday         44373         28.980236
0     Friday         44373         32.991526
2   Saturday         44373         28.667455
3     Sunday         44373         30.500169


### Question 3c

In [105]:
# I will be using the mean as the threshold
# this seperates days with above-average/median visits,
# with days which are below that
threshold = df_long['dailyvisits'].mean()

# Define 'manyvisits' as 1 if dailyvisits exceeds the threshold, otherwise 0
df_long['manyvisits'] = (df_long['dailyvisits'] > threshold).astype(int)

# Create a pivot table
pivot_table = (df_long.pivot_table(index='region', 
                columns='manyvisits', aggfunc='size', fill_value=0))

# Calculate the total observations and the percentage of 'manyvisits=1'
pivot_table['total_observations'] = pivot_table[0] + pivot_table[1]
pivot_table['percentage_manyvisits_1'] = ((pivot_table[1] 
                                           / pivot_table['total_observations'])
                                          * 100)

# Sort the table by 'percentage_manyvisits_1' 
# and then by the count of 'manyvisits=1', both in descending order
sorted_pivot_table = (pivot_table.sort_values
                      (by=['percentage_manyvisits_1', 1],
                       ascending=[False, False]))

# Identify the geographic unit(region) with the highest number of 'manyvisits=1'
highest_manyvisits = sorted_pivot_table.iloc[0][1]

# Identify the geographic unit with the highest percentage of 'manyvisits=1'
highest_percentage = sorted_pivot_table['percentage_manyvisits_1'].idxmax()

print(f"Geographic unit with the highest number of 'manyvisits=1': {highest_manyvisits}")
print(f"Geographic unit with the highest percentage of 'manyvisits=1': {highest_percentage}")

# Apply styling
styled_table = (sorted_pivot_table.style.background_gradient
                (cmap='coolwarm', subset='percentage_manyvisits_1')\
    .format("{:.0f}", subset=[0, 1, 'total_observations'])\
    .format("{:.2f}%", subset='percentage_manyvisits_1')\
    .set_caption("Geographic Summary with Manyvisits")\
    .bar(subset=[1, 'percentage_manyvisits_1'], color='#5fba7d'))

styled_table


Geographic unit with the highest number of 'manyvisits=1': 9049.0
Geographic unit with the highest percentage of 'manyvisits=1': MS


Zaxby's is most popular in the South, which you can see with the top most popular states. The public information support my findings.

### Question 3d

In [106]:
# Number of units per region 
units_per_region = df_long['region'].value_counts()

# Threshold - the 80th percentile of units_per_region
threshold_units = units_per_region.quantile(0.7)

# Core business areas
df_long['core_biz_area'] = (df_long['region'].apply
                            (lambda x: 1 if units_per_region[x] 
                             >= threshold_units else 0))
# Aggregate data by core_biz_area
aggregation = df_long.groupby('core_biz_area').agg(
    number_of_observations=('dailyvisits', 'size'),
    mean_dailyvisits=('dailyvisits', 'mean')
).reset_index()

print(aggregation)


   core_biz_area  number_of_observations  mean_dailyvisits
0              0                  128457         29.598317
1              1                  182154         28.285001


To my surprise, I found that the businesses NOT in my core business area produced a high mean of daily visits. This may represent an untapped market with less competition and high demand. Since there are fewer units in a single area, those units may be benefiting from more operational focus, enhancing customer experiences and attractign more visits. But, just to be sure, I want to approach the analysis a different way and rethink what I define as a core business area. In the approach below, I will be looking at areas where my mean daily visits are in the top quartile, and making those my core business areas. 

In [107]:
# Calculate mean daily visits per region
region_performance = df_long.groupby('region')['dailyvisits'].mean()

# Define core areas based on performance - 75% is top quartile
performance_threshold = region_performance.quantile(0.75)
df_long['core_biz_area_by_performance'] = (df_long['region'].
                                           apply(lambda x: 1 
                                          if region_performance[x] 
                                          >= performance_threshold else 0))

# Aggregate data by the new core_biz_area definition
aggregation_by_performance = df_long.groupby('core_biz_area_by_performance').agg(
    number_of_observations=('dailyvisits', 'size'),
    mean_dailyvisits=('dailyvisits', 'mean')
).reset_index()

print(aggregation_by_performance)


   core_biz_area_by_performance  number_of_observations  mean_dailyvisits
0                             0                  173971         22.721057
1                             1                  136640         36.603718


Now I see that I have a higher daily visit in my core area.  